cfg/CFG.h

```diff
@@ -120,26 +125,48 @@ class CFG final {
     static constexpr int ENSURE_BLOCK_OFFSET = 2;
     static constexpr int ELSE_BLOCK_OFFSET = 3;
 
-    UnorderedMap<core::LocalVariable, int> minLoops;
-    UnorderedMap<core::LocalVariable, int> maxLoopWrite;
-
     void sanityCheck(core::Context ctx);
 
     struct ReadsAndWrites {
-        std::vector<UnorderedSet<core::LocalVariable>> reads;
-        std::vector<UnorderedSet<core::LocalVariable>> writes;
+        std::vector<UnorderedSet<int>> readsSet;
+        std::vector<std::vector<int>> reads;
+        std::vector<std::vector<int>> writes;
 
         // The "dead" set reports, for each block, variables that are *only*
         // read in that block after being written; they are thus dead on entry,
         // which we take advantage of when building dataflow information for
         // inference.
-        std::vector<UnorderedSet<core::LocalVariable>> dead;
+        std::vector<std::vector<int>> dead;
     };
     ReadsAndWrites findAllReadsAndWrites(core::Context ctx);
+    LocalRef enterLocal(core::LocalVariable variable);
```

Comment:
This API is public, _but_ we enforce via freezing that only privileged APIs can control when it's valid to define new locals, as `maxVariableId` is used to size several vectors later on in CFG.